src/node_api.cc

```diff
@@ -491,15 +476,42 @@ class TryCatch : public v8::TryCatch {
 
 //=== Function napi_callback wrapper =================================
 
-static const int kDataIndex = 0;
-static const int kEnvIndex = 1;
+static const int kFunctionIndex = 0;  // Used in CallbackBundle::cb[]
+static const int kGetterIndex = 0;    // Used in CallbackBundle::cb[]
+static const int kSetterIndex = 1;    // Used in CallbackBundle::cb[]
+static const int kCallbackCount = 2;  // Used in CallbackBundle::cb[]
```

Comment:
Is there a reason to use indices like this instead of separate fields on `CallbackBundle`?

Reply:
Good question, the indices are there for a historical reason (also be used by `GetInternalField()` function)... now they don't have to be indices... but changing this would spread the change to some template classes (`XXXCallbackWrapper` x3). I didn't do this to keep the change set minimal and reasonable (IMHO).

Reply:
I’d appreciate it if you could either change it here (no need to aim for minimality in that respect imo), or alternatively leave a `// TODO(somebody)` comment to remind us that we want to do this soon afterwards then

Reply:
Actually, the indices are useful in `InvokeCallback()` which is templated on the index value. I'm not sure if we can template on a field name.

Reply:
Yes, removing the indices would require 3 new "template traits classes" to be invented, which are supposed to be used in `CallbackWrapperBase<>::InvokeCallback()` to extract the correct data member of `CallbackBundle` (the data members of `CallbackBundle` are likely to be organized in union to save memory)... but what's the benefit of doing so? Not so obvious...

Of course we can introduce a bigger change to discard the `CallbackWrapperBase<>` template to avoid writing new traits classes, but is it worth it? A fixed dimension array and well-controlled indices are simple and "just works" (minimizing the cognitive load)

Just my two cents :)

Reply:
> I'm not sure if we can template on a field name.

That is possible in C++, yes:

```c++
struct Foo {
  char field1;
  char field2;
  char field3;
};

template <char (Foo::* Field)>
class FieldGetter {
 public:
  static char Get(Foo* foo) {
    return foo->*Field;
  }
};

char Example() {
  struct Foo a = { 1, 2, 3 };

  return FieldGetter<&Foo::field2>::Get(&a);
}
```

It’s just a stylistic thing so it’s very fine to leave it out here, but as I said above, it would be nice to leave a `TODO` comment in that case

Reply:
TODO comment added in the most recent push.